### PR TITLE
Feature/6742 vaccination certificate/6940 decoding

### PR DESCRIPTION
**Pull request by ArturFriesen**
## Description
Add functionality to decode base45 to header and certificate representations.

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6940

### CODE REVIEW

**Review by 30mar — APPROVED** [review]

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/CertificateAccess.swift [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            case let .failure(error):
+                return .failure(error)
+            }
+
+        case let .failure(error):
+            return .failure(error)
+        }
+    }
+
+    func extractHealthCertificate(from cborWebToken: CBOR) -> Result<DigitalGreenCertificate, HealthCertificateDecodingError> {
```
[the remark this answers is not in the document]
Comment (nickguendling) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/CertificateAccess.swift:77]: I guess this applies also to all healthCertificate naming inside this func

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/CBORWebTokenHeader.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+        case issuer = "iss"
+        case issuedAt = "iat"
+        case expirationTime = "exp"
+    }
+
+    // MARK: - Internal
+
+    let issuer: String
+    let issuedAt: UInt64?
+    let expirationTime: UInt64
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/CBORWebTokenHeader.swift:21]: Could these be `Date`s? and be named `issueDate` and `expirationDate`? But we could also convert that on the Service level / introduce computed properties for that I guess 🤔

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/Error.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public enum HealthCertificateDecodingError: Error {
+    case HC_BASE45_DECODING_FAILED
+    case HC_ZLIB_DECOMPRESSION_FAILED
+    case HC_COSE_TAG_INVALID
+    case HC_COSE_MESSAGE_INVALID
+    case HC_CBOR_DECODING_FAILED
+    case HC_CWT_NO_ISS
+    case HC_CWT_NO_EXP
+    case HC_CWT_NO_HCERT
+    case HC_CWT_NO_DGC
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/Error.swift:21]: As mentioned earlier I would prefer names that are easier to read and add the original names from the tech spec as a comment

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/ProofCertificateAccess.swift [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+public struct ProofCertificateAccess {
+
+    // MARK: - Public
+
+    public func extractCBORWebTokenHeader(cbor: Data) -> Result<CBORWebTokenHeader, HealthCertificateDecodingError> {
+        return certificateAccess.extractHeader(from: cbor)
+    }
+
+    public func extractHealthCertificate(cbor: Data) -> Result<DigitalGreenCertificate, HealthCertificateDecodingError> {
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/ProofCertificateAccess.swift:15]:
```suggestion
    public func extractDigitalGreenCertificate(cbor: Data) -> Result<DigitalGreenCertificate, HealthCertificateDecodingError> {
```

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/CBOR+AnyMap.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+            return dateValue as Any
+        case .utf8String(let stringValue):
+            return stringValue as Any
+        case .array(let arrayValue):
+            return arrayValue.map { $0.anyValue }
+        case .map(let mapValue):
+            return mapValue.anyMap as Any
+        case .null, .undefined:
+            return nil as Any?
+        default:
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/CBOR+AnyMap.swift:51]: Is the default required? Are there more cases here that could be handled or state explicitly?

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import Foundation
+import Compression
+
+extension Data {
+
+    func decompressZLib() throws -> Data {
+
+        // The maximum output size of the zlib decompression shall be set to 10 MB to protect against zip bomb attacks.
+        let tenMBCapacityLimitInByte = 10_000_000
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift:13]: Is 10 * 1000 * 1000 correct or should it be 10 * 1024 * 1024 (10_485_760)?

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+            guard let baseAddress = typedPointer.baseAddress else {
+                throw HealthCertificateDecodingError.HC_ZLIB_DECOMPRESSION_FAILED
+            }
+
+            let read = compression_decode_buffer(
+                decodedDestinationBuffer,
+                tenMBCapacityLimitInByte,
+                baseAddress,
+                count - 2, nil,
```
Comment (nickguendling) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift:29]:
```suggestion
                count - 2, 
                nil,
```

**Review by nickguendling — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift (hunk 6 above)
Comment (ArturFriesen) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift:13]: Right. I tried to sell you a hard drive. You got me.

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+//
+
+import Foundation
+import base45_swift
+
+public struct HealthCertificateAccess {
+
+    // MARK: - Public
+
+    public func extractCBORWebTokenHeader(base45: String) -> Result<CBORWebTokenHeader, HealthCertificateDecodingError> {
```
Comment (whiskey) [inline src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift]: > base45: String

I would suggest a `typealias Base45 = String` to distinguish encoded strings from normal ones.

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/CBOR+AnyMap.swift (hunk 5 above)
Comment (ArturFriesen) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/CBOR+AnyMap.swift:51]: added missing cases .tagged, .break and removed default

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift (hunk 8 above)
Comment (whiskey) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift]: Similar to the CBOR data (and potentially some other types in the app as well): `typealias CBORData = Data`

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/Error.swift (hunk 3 above)
Comment (ArturFriesen) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Model/Error.swift:21]: done

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift (hunk 6 above)
Comment (whiskey) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/Utils/Data+Compression.swift]: Wouldn't that be 'BiByte' then? 🤷‍♂️

**Review by nickguendling — APPROVED** [review]

File: src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift (hunk 8 above)
Comment (ArturFriesen) [reply src/xcode/ENA/HealthCertificateToolkit/Sources/HealthCertificateToolkit/HealthCertificateAccess.swift]: 🤩 Like it! Added!

**Review by whiskey — APPROVED** [review]

**Pull request merged**